[WIP] Added toWarnInDev matcher and updated one test to use it

bvaughn opened the pull request:
Relates to #11626. Pushing for discussion purposes. 😄 

Open questions:
* Is this worth doing?
* Is the automatic spy-on behavior too much of a hack?
* Is this method of differentiating between low-priority (`console.warn`) and high-priority (`console.error`) warnings ok?
* Is there a way to combine `.toThrowError` and `.toWarnDev` matchers (see "_includes the owner name when passing null, undefined, boolean, or number_")?

thymikee commented on scripts/jest/matchers/toWarnDev.js:
```diff
@@ -0,0 +1,102 @@
+  if (__DEV__) {
+    if (typeof expectedWarnings === 'string') {
+      expectedWarnings = [expectedWarnings];
+    } else if (!Array.isArray(expectedWarnings)) {
```
@bvaughn Are you planning on using Flow inside tests at some point? If so, you could get rid of these checks.

bvaughn replied:
True.

And no idea. 😄

gaearon commented on packages/react/src/__tests__/ReactElementValidator-test.js:
```diff
@@ -248,61 +214,48 @@ describe('ReactElementValidator', () => {
+    ]);
+
+    // Should not log any additional warnings
+    expect(() => React.createElement('div')).toWarnDev([]);
```
Is this necessary? Shouldn’t an unexpected log fail the test anyway? I expect that if we explicitly mark each warning block then any warning outside should be a failure. Maybe even immediately when it’s called rather than at the end of the test.

bvaughn replied:
> Shouldn’t an unexpected log fail the test anyway

Good point.

bvaughn replied:
Now that I look at this again, I believe this explicit check is necessary since we've necessarily spied on `console.error` in order for the matcher to work.

IIR, I originally tried to clear the spy at the end of the matcher, so unexpected errors/warnings would cause an error, but neither the version of Jest we're using (`21.3.0-beta.4`) nor the latest release (22) actually expose the `spy.mockReset` method [the docs mention](http://facebook.github.io/jest/docs/en/jest-object.html#jestspyonobject-methodname).

We could call `spy.and.callThrough()` at the bounds of the matcher, so that errors were actually passed through, but (a) that could negatively affect tests with multiple `.toWarnDev` calls and (b) this wouldn't actually fail the test, only print the red text to the console.

Maybe I'm overlooking something?

gaearon replied:
>Now that I look at this again, I believe this explicit check is necessary since we've necessarily spied on console.error in order for the matcher to work.

Wouldn't this approach solve it? https://github.com/facebook/react/pull/11786#discussion_r155672766

>IIR, I originally tried to clear the spy at the end of the matcher, so unexpected errors/warnings would cause an error

I don't quite understand what clearing means in this context, or why it would be helpful.

My proposal would be to set up a mock when we enter `toWarn` and tear it down when we exit it. Outside of `toWarn`, I'd want `console.error` to be replaced with a function that insta-throws.

bvaughn replied:
I think we're saying the same thing, but misunderstanding each other.

My initial design for this matcher was to setup a spy/mock when entered and then remove it before existing. This way unexpected `console.error` calls would fail the test (as they do if you don't spy in the first place). The Jest docs [suggest this should be possible](http://facebook.github.io/jest/docs/en/jest-object.html#jestspyonobject-methodname) via a `mockReset` method, but that method doesn't actually exist (in Jest 21 or 22) so I'm not sure why the docs mention it. The problem then, is that once we've spied on `console.error` initially, we can't remove that spy and so any unexpected calls to it don't actually cause failures.

We could work around this in other ways. For example, I could not use Jest's `spyOn` functionality at all- but just override `console.error` with my own function and then restore the originally one before exiting. Maybe that's best anyway?

gaearon replied:
>We could work around this in other ways. For example, I could not use Jest's spyOn functionality at all- but just override console.error with my own function and then restore the originally one before exiting. Maybe that's best anyway?

Sorry I wasn't clear; yes, this is what I meant.

gaearon replied:
>My initial design for this matcher was to setup a spy/mock when entering and then remove it before exiting. This way unexpected console.error calls would fail the test (as they do if you don't spy in the first place)

They still wouldn't fail it *immediately*. Only upon checking the mock (which currently happens at the end of the test). But this is annoying because you have to "backtrack" to see *where exactly* an extra warning has fired and why.

My proposal was to override these methods so we can *replace them with shims that throw* when we're outside of the `toWarn` block. This way calling them would fail the test *immediately* rather than we have reached the end of the test. But it does require us to roll our own.

bvaughn replied:
I think that's actually a reasonable plan for this matcher, since we don't need full the spy functionality as implemented by Jasmine/Jest anyway.

bvaughn replied:
Oh sorry, looks like our comments got crossed.

WRT just rolling our own "spy"- cool. I'll do this.

Regarding the second thing you say, I think you're just suggesting we modify our `console` overrides [here](https://github.com/facebook/react/blob/247c5241706356033cd5bf1dc31197e058853e13/scripts/jest/setupTests.js#L58-L87) to throw immediately rather than in an `afterEach`? I don't think this is specifically related to this PR, but I'd be happy to roll it into this one. 😄

gaearon replied:
>Regarding the second thing you say, I think you're just suggesting we modify our console overrides here to throw immediately rather than in an afterEach? I don't think this is specifically related to this PR, but I'd be happy to make that change as well.

It's related to making this PR more than an aesthetic change :-) I think throwing early extracts a benefit from the new API—something we haven't been able to do before. It also prevents accidental use of old API (directly asserting on the mock). But I agree it's not a prerequisite.

bvaughn replied:
Hmm, ok. Although I think aesthetics/ergonomics are important for test writing.

> I think throwing early extracts a benefit from the new API—something we haven't been able to do before.

Sorry if I'm being dense, but I don't understand what you mean about us being unable to throw immediately before. Could you explain?

We always wanted unexpected warnings to fail, and we always signified they were expected by spying on them. That hasn't changed with this PR.

bvaughn replied:
Never mind. I believe I understand now.

gaearon replied:
With this API:

```js
spyOnDev(console, 'warn'); // line 1

console.warn('good warning'); // line 2
console.warn('unexpected warning'); // line 3
doSomethingElse(); // line 4

expectDev(console.warn.calls.length).toBe(1); // line 5
```

if there is a second (unexpected) warning, the test fails on line 5. You have to hunt `console.error` calls to see that the unexpected warning was caused by line 3 (in practice, it's somewhere inside the library).

With this API:

```js
expect(() => { // line 1
  console.warn('good warning'); // line 2
  console.warn('unexpected warning'); // line 3
}).toWarnInDev(['stuff']) // line 4

console.warn('stuff'); // line 5
doSomethingElse(); // line 6
```

the test fails directly on line 5, with a callstack pointing to the `console.warn` call. We don't need to wait for the whole section to execute (or to check the expected warning count twice) to know it has failed. And we have the exact stack trace for *where* the unexpected warning happened.

-----

Similarly, another scenario before and after:

```js
doSomething();
console.warn('did not expect this warning');
doSomethingElse();
```

In our current setup this test fails *after* it has finished executing. We have no clue as to *where in the code* the warning has happened.

Compared to the same code:

```js
doSomething();
console.warn('did not expect this warning');
doSomethingElse();
```

Since we now know that any warning outside of `toWarn` is bad, we can have it throw immediately, and thus show the stack trace.

---

If this still doesn't make sense, never mind, maybe I'm not explaining it well and it's not that important anyway. All I'm saying is that `toWarn()` kind of API gives us the ability to know where exactly the "expected zone" for warnings starts and ends, as opposed to the previous API whose granularity was "something in this test warns somewhere". So previously we couldn't fail immediately on a warning outside of that zone since the whole test was "fair game". But now we know exactly which lines can warn, and if any line outside warns, we can fail.

In fact now we can validate the warning messages immediately.

With

```js
expect(() => {
  console.error('unexpected stuff');
  doSomething();
  console.error('stuff');
  console.error('other stuff');
}).toWarn([
  'stuff',
  'other stuff'
]);
```

we can now throw right in the first `console.error` call because *we already know it has mismatched* and there's no point executing the test further (and obscuring what happened by letting variables be mutated, etc).

On the other hand, with the current approach

```js
spyOnDev(console, 'error');

console.error('unexpected stuff');
doSomething();
console.error('stuff');
console.error('other stuff');

expect(console.errors.calls.length).toBe(2);
```

we only know something went wrong by the time we reached the last line, and the state might have been obscured by now. Usually I end up having to comment out every line one by one until I find the earliest one where the mismatched warning still happens.

gaearon replied:
(My initial comment didn't make sense so I edited it)

bvaughn replied:
Ah! I think I understand what you're saying better now.

The boundary is clearer with the new API. Before the boundary was more informal, between `spyOn` and the first `(...calls.length).toBe()` check.

It was still possible to do the throw-immediate thing before though. That's what confused me a bit.

bvaughn replied:
I guess it was also easier for the old spies to miss things if warnings were logged after the last count check too. That's another advantage of this boundary.

gaearon replied:
>It was still possible to do the throw-immediate thing before though

How would it work with this test case?

```js
spyOnDev(console, 'error')

console.error('unexpected warning')
console.error('expected warning')

expect(console.error.calls.count).toBe(1)
expect(console.error.calls.argsFor(0)[0]).toBe('expected warning')
```

Previously we couldn't safely throw on the first `console.error` because we didn't know its message is unexpected by the time it ran.

But with `toWarn` we can because we know the exhaustive list of expected messages (and their count) *before* `console.error` is called.

bvaughn replied:
It just occurred to me coming back from lunch that, since I know the error message(s) ahead of time, I could also throw immediately (within my fake spy) any time a message is logged that isn't expected.

I'll rework the matcher to behave this way instead.

gaearon commented on packages/react/src/__tests__/ReactElementValidator-test.js:
```diff
@@ -317,6 +270,8 @@ describe('ReactElementValidator', () => {
     );
+
+    // We can't use .toWarnDev() here because we can't chain it with .toThrowError()
     if (__DEV__) {
```
Let’s bite the bullet and nest assertions?

bvaughn replied:
I'm not sure what you have in mind here. Mind giving a faux code example?

gaearon replied:
```js
expect(() => {
  expect(() => {
  
  }).toWarnInDev()
}).toThrowError('message')
```

Although there's a hazard of not specifying the message and thus catching a failure from the inner expect.

gaearon replied:
Maybe nesting them the other way around would work.

bvaughn replied:
Nesting `toWarnDev` inside of `toThrowError` won't work. Errors in `toWarnDev` are ignored. But nesting it the other way around (`toThrowError` inside) _does_ seem to work.

gaearon commented on packages/react/src/__tests__/ReactElementValidator-test.js:
```diff
@@ -388,36 +331,28 @@ describe('ReactElementValidator', () => {
+    ]);
+
+    // Should not error for strings
+    expect(() => {
```
Same; I would expect any warning that isn’t inside a protected block to fail immediately.

bvaughn replied:
I think this was just me being silly.

gaearon replied:
To be clear, I expect that with current infra it fails at the end of the test.

However, I propose that `toWarnInDev` acts as a boundary, and we change `console.error`/`console.warn` to **immediately throw** outside of that boundary. This way we get the error immediately at the right call site instead of at the end of the test.

bvaughn commented:
I've refactored this to fail faster for unexpected warnings. I also added some syntax snippets to the description.

Still need to look into the following:
- [ ] Do we actually need to support nested warnings? (Probably not.) If not, then we can just error on this case.
- [ ] Is there a way to get a better error stack for unexpected warnings? Looks like Jest is catching truncating it to just the test where the expectation is defined. I could add the stack to the thrown error message but that seems a little hacky.

bvaughn commented:
It looks like the [stack thing I mentioned is intentional](https://github.com/nicolasiensen/jest/blob/c0d78a92591fbef075d504a75759cfd49beecffb/packages/expect/src/index.js#L200-L208). Jest overrides the "real" stack with one that points to the matcher.

bvaughn commented:
I filed a Jest issue (facebook/jest/issues/5136) and PR (facebook/jest/pull/5138) for preserving the call stack for errors inside of a custom matcher. I don't think either is a blocker for moving forward with this issue though, so back to you @gaearon 😄

gaearon commented:
What did we decide about throwing?

We should make sure this fails the test:

```js
try {
  console.error('hi');
} catch (err) { }
```

bvaughn commented:
That's a good point, Dan. Using the `afterEach` to throw, like we did before, obscured the call stack. Throwing synchronously might get swallowed up and suppressed though.

I think the best compromise might be to track call stacks during execution, save them for later, then throw a single error from the `afterEach` block. This would not prevent the test from continuing to execute (unfortunately, I don't know a way to reliably do that) but it would hopefully simplify the debugging/test-authoring experience.

Maybe I'm overdoing this, but here's a compromise that I think might work? For example, this test:

```js
describe('Example', () => {
  it('should warn about an unexpected log', () => {
    const foo = () => bar();
    const bar = () => baz();
    const baz = () => console.error('unexpected warning');

    foo();
  })

  it('should warn about two unexpected logs', () => {
    const foo = () => bar();
    const bar = () => baz();
    const baz = () => console.error('unexpected warning');

    try {
      foo();
      console.error('another unexpected warning')
    } catch (err) { }
  })
});
```

Would print the following failures:
![screen shot 2017-12-19 at 2 47 36 pm](https://user-images.githubusercontent.com/29597/34182539-946886e2-e4cb-11e7-889f-822af15ba997.png)

bvaughn commented:
I've pushed a fix for this (as described above) and updated the description of this PR for the various cases I've tested (locally) with this matcher.

gaearon commented:
I think this makes sense. I'll review this tomorrow.

bvaughn commented:
Friendly ping 😄

bvaughn commented:
Another friendly ping

gaearon commented:
Sorry, I was on holidays!

bvaughn commented:
Of course! 😄  Was just a reminder for when we both came back.

gaearon commented on scripts/jest/matchers/toWarnDev.js:
```diff
@@ -0,0 +1,80 @@
+        }
+
+        // Fail early for unexpected warnings to preserve the call stack.
+        throw Error(`Unexpected warning recorded: "${message}"`);
```
Can we include the *expected* warning message if one exists, in this message?

Otherwise it's a bit confusing that simply having a typo in the expected warning prints a message as if a warning was not expected at all.

bvaughn replied:
Yeah. I like this suggestion.

What do you think about the following change?
![screen shot 2018-01-02 at 10 50 01 am](https://user-images.githubusercontent.com/29597/34495593-b91371de-efaa-11e7-8dce-1196d37ea62f.png)

bvaughn replied:
To be clearer:
```js
it('should fail when the expected warnings passed to .toWarnDev() do not match the actual warnings', () => {
  expect(() => {
    console.error('foo');
  }).toWarnDev('bar');
});
```
![screen shot 2018-01-02 at 10 58 04 am](https://user-images.githubusercontent.com/29597/34495846-dae6bdc4-efab-11e7-9fd7-f654c294aa62.png)

DavidCurtis321 commented:
Get Outlook for iOS
	





On Tue, Jan 2, 2018 at 8:08 AM -0500, "Dan Abramov" <notifications@github.com> wrote:












Sorry, I was on holidays!



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.

bvaughn commented:
Thanks, but I don't need Outlook 😉

gaearon commented on scripts/jest/setupTests.js:
```diff
@@ -56,16 +62,22 @@ if (process.env.REACT_CLASS_EQUIVALENCE_TEST) {
 
     env.beforeEach(() => {
-      newMethod.__callCount = 0;
+      unexpectedConsoleCallStacks.splice(0);
```
I always forget what this does. Maybe ` = []` or `.length = 0`?

gaearon commented:
How do you feel about converting other tests as part of this PR? I don't care strongly either way.

bvaughn commented:
I could do that, although this PR is already pretty big. I stopped converting because I was getting increasing concerned about merge conflicts. 😄  I'd be happy to do a follow up PR or two quickly after merging this.

I'm willing to do it all as part of this PR too though if you have a preference.

gaearon commented:
Let's split (as long as both styles keep working and there's no regressions).
But we'll have merge conflicts anyway (since it's likely your PRs will land before any others)/

bvaughn commented:
True. I was mostly worried about having to deal with merge conflicts _myself_ 😅  I'll cause potential conflicts for others no matter what I do.

I've merged `master` into this branch and addressed your 2 comments above. Once CI passes, I'll merge this and then continue updating other tests with a follow-up PR.

gaearon commented:
As a stretch goal it’d be nice to highlight the text difference on a character level. I often make small tweaks and then it’s super annoying to stare into the message trying to figure out where exactly the different lies (and which one is right).

bvaughn commented:
Yeah, I briefly considered doing this, using `jest-diff`, but the differences are too big. For examples, there's a "Warning:" prefix that may not be there, and there's the component stack that may be different - neither of which are actually important in the context of the test in many cases.

Maybe we could do some kind of minimum-edit-distance calculation to narrow down the area to diff?
